### PR TITLE
Individualized run_every period for each rule

### PR DESCRIPTION
**Pull request by mircopolo**
Changes to allow rules to have individualized run_every periods. Still very much in beta needs much review for thread safety etc, plus figure out how to hand reloaded rules who's name have changed etc.. 

Main changes are as follows
1) Update config code to load run_every from rule_file.yaml if present 
2) Replace certain EA instance attributes with thread_local storage (counters and es_clients)
3) Split run_all into separate functions for sending pending alerts, reloading config and executing a single rule instance. 
4) Added APScheduler to schedule each rule in its own job, two further background threads for pending alerts and config_reloads.

This PR should work very nicely with the cron syntax changes in #1114, but at the moment just using a straight up interval scheduling.

### CODE REVIEW

**Comment by mircopolo** [comment]
Should address #497, #644

File: elastalert/elastalert.py [hunk 1]
```diff
@@ -575,7 +574,7 @@ def run_query(self, rule, start=None, end=None, scroll=False):
                 rule_inst.add_data(data)
 
         try:
-            if rule.get('scroll_id') and self.num_hits < self.total_hits:
+            if rule.get('scroll_id') and self.thread_data.num_hits < self.total_hits:
```
Comment (Qmando) [inline elastalert/elastalert.py]: I think total_hits need to be in thread_data too.

File: elastalert/elastalert.py [hunk 2]
```diff
@@ -875,7 +876,9 @@ def init_rule(self, new_rule, new=True):
             if prop not in rule:
                 continue
             new_rule[prop] = rule[prop]
-
+                   
+        self.scheduler.add_job(self.handle_rule_execution, 'interval', args=[new_rule] ,seconds=new_rule['run_every'].total_seconds(),id=new_rule['name'])
```
Comment (Qmando) [inline elastalert/elastalert.py]: Style: Add spaces after commas

Also, lets add max_instances=1, which should prevent overlapping runs if execution time goes over run_every

File: elastalert/elastalert.py [hunk 3]
```diff
@@ -992,53 +1002,64 @@ def start(self):
...
+    
+    def handle_config_change(self):
         if not self.args.pin_rules:
             self.load_rule_changes()
-
+            elastalert_logger.info("Background configuration change check run at %s" % (pretty_ts(ts_now())))
+    
+    def handle_rule_execution(self, rule):
+      self.thread_data.alerts_sent = 0
+      next_run = datetime.datetime.utcnow() + self.run_every
```
Comment (Qmando) [inline elastalert/elastalert.py]: rule['run_every'] ?

**Comment by kvalev** [comment]
I can also help out with the validation when you think that the pull request is ready to be tested.

**Comment by Qmando** [comment]
So you know, I haven't given up on this PR. I've made some modifcations to it and opened a new pull request, #1298

**Comment by mircopolo** [comment]
Thanks @Qmando , I actually missed the notification for the code review so didn't realise it had been looked at recently. I had meant to come back to complete this at some point but been pretty busy for the last while, but looks like you have everything under control. 

I'll try and pull down the other new PR to help with some testing if I can.

**Pull request closed without merging**